added functions to estimate incubation period and docs

ffinger opened the pull request:
WP3: incubation period for ambiguous dates of exposure from gdoc. vignette with doc provided.

ffinger committed:
added functions to estimate incubation period and docs

ffinger committed:
spelling in doc

ffinger committed:
plot in vignette

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+fit_gamma_incubation_dist <- function(x, dates_exposure, date_of_onset, nsamples = 1000, ...) {
+  incubation_period_dist <- empirical_incubation_dist(x, exposures, date_of_onset)
+
+  s <- base::sample(
```
You may want to put in a catch here to make sure that the `incubation_period_dist` contains more than one row of data. Otherwise, `base::sample()` will assume that you want to sample from 1 to whatever integer you provide.

``` r
sample(c(9, 10), 5, replace = TRUE)
#> [1] 10 10 10 10  9
sample(c(10), 5, replace = TRUE)
#> [1] 8 7 3 1 9
```

<sup>Created on 2019-02-07 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

zkamvar commented on vignettes/estimate_incubation.Rmd:
```diff
@@ -0,0 +1,84 @@
+     toc: true
+     toc_depth: 2
+vignette: >
+  %\VignetteIndexEntry{estimate_incubation}
```
Thank you for actually putting in a title for the index entry 🙌 

That being said, the entry can have spaces, so could you change this to "Estimating Incubation Period"?

zkamvar commented on vignettes/estimate_incubation.Rmd:
````diff
@@ -0,0 +1,84 @@
+Load environment:
+
+```{r, echo = TRUE, message=FALSE}
+library(linelist)
````
Please add linelist to the `Suggests` field in the `DESCRIPTION` file. Also, add the following:

```
Remotes:
  reconhub/linelist
```

zkamvar commented on vignettes/estimate_incubation.Rmd:
```diff
@@ -0,0 +1,84 @@
+case_types <- c("confirmed", "probable", "suspected", "not a case",
+                "Confirmed", "PROBABLE", "suspected  ", "Not.a.Case")
+case <- factor(sample(case_types, 20, replace = TRUE))
+toy_data <- data.frame("Date of Onset." = onsets,
```
you can use `linelist::messy_data()` to create a random messy data set

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+#' @author Flavio Finger, \email{flavio.finger@lshtm.ac.uk}
+#' @export
+fit_gamma_incubation_dist <- function(x, dates_exposure, date_of_onset, nsamples = 1000, ...) {
+  incubation_period_dist <- empirical_incubation_dist(x, exposures, date_of_onset)
```
Should `exposures` be `dates_exposure`? 

Additionally, I believe you also have to escape these for tidyeval to work:

```r
  library(dplyr)
  library(rlang)
  headselect <- function(x, y) {
    head(dplyr::select(x, !!rlang::enquo(y)))
  }
  wrapper <- function(x, y) {
    headselect(x, y)
  }
  wrapper2 <- function(x, y) {
    headselect(x, !!rlang::enquo(y))
  }
  headselect(iris, Petal.Length) # works
  wrapper2(iris, Petal.Length)   # works
  wrapper(iris, Petal.Length)    # error
```

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+#' Can take into account uncertain dates of exposure.
+#'
+#' @param x the linelist data (data.frame or linelist object) containing at least a column containing the exposure dates and one containing the onset dates. For exposure dates, each element can be a list containing several possible exposure dates.
+#' @param dates_exposure the name of the column containing the exposure dates.
```
Please indicate here that the variables are bare names and not strings.

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+#' @param date_onset the name of the column containing the onset dates
+#' @return a data frame containing a column with the different incubation periods and a column containing their relative frequency
+#' @author Flavio Finger, \email{flavio.finger@lshtm.ac.uk}
+#' @export
```
Please add a short example here

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+  dates_exposure <- rlang::enquo(dates_exposure)
+  date_of_onset <- rlang::enquo(date_of_onset)
+
+  y <- x %>%
```
I think this whole block needs to be moved into a separate internal function that handles two date vectors. That way, it will be easier to debug later on.

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+      incubation_period = as.numeric(!!date_of_onset - !!dates_exposure),
+      weight = as.numeric(weight)
+    ) %>%
+    dplyr::select(incubation_period, weight) %>%
```
R will throw a note saying that `incubation_period` and `weight` are undefined here

ffinger replied:
I don't get why? They are columns defined within mutate...

zkamvar replied:
Even though they are defined in mutate, they aren't defined in the scope of the function and even though R knows what to do in this situation, it feels uncomfortable about it. More extensive reading on the situation: https://stackoverflow.com/q/9439256/2752888

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+  date_of_onset <- rlang::enquo(date_of_onset)
+
+  y <- x %>%
+    dplyr::select(!!dates_exposure, !!date_of_onset) %>%
```
All of the dplyr/rlang/magrittr functions need to be explicitly imported using roxygen tags (that being said, only !! and %>% really NEED imports)

```r
#' @importFrom dplyr mutate unnest select
#' @importFrom rlang "!!" enquo
#' @importFrom purrr map
#' @importFrom tidyr unnest
#' @importFrom magrittr "%>%"
```

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,67 @@
+#' @param ... passed to fit_disc_gamma
+#' @return see fit_disc_gamma
+#' @author Flavio Finger, \email{flavio.finger@lshtm.ac.uk}
+#' @export
```
Please add a small example

zkamvar commented on vignettes/estimate_incubation.Rmd:
````diff
@@ -0,0 +1,84 @@
+
+```{r, echo = TRUE, message=FALSE}
+library(linelist)
+library(dplyr)
````
For the vignette, you don't need to load all these packages, just linelist, epitrix, magrittr, and ggplot2

zkamvar commented on vignettes/estimate_incubation.Rmd:
```diff
@@ -0,0 +1,84 @@
+fit <- fit_gamma_incubation_dist(x, exposures, date_of_onset)
+print(fit)
+
+ggplot(transform(data.frame(x=c(0:10)), y=fit$distribution$d(x)), aes(x, y)) +
```
What does transform do here and what package is it from?

zkamvar commented on vignettes/estimate_incubation.Rmd:
```diff
@@ -0,0 +1,84 @@
+
+x <- toy_data %>%
+  clean_data %>%
+  as_linelist(date_onset = "date_of_onset")
```
Currently, nothing is really being done with `date_of_onset` here...

ffinger replied:
yes, I thought of maybe making it possible to use the epivars in linelist to define the columns, but haven't (yet) implemented that

zkamvar reviewed (changes requested):
Thank you for contributing, @ffinger!

I think this is a good start! It's currently [not passing check on Travis](https://travis-ci.org/reconhub/epitrix/jobs/489667134#L1314) primarily because it can't find linelist, but it will also fail if you fix that. To get this working, you'll need to do the following in the `DESCRIPTION`

1. add linelist to Suggests
2. add the following lines to the DESCRIPTION: 
    ```
    Remotes:
      reconhub/linelist
    ```
2. add dplyr, purrr, magrittr, and tidyr to the Imports

I'm still on the fence about whether or not it's a good idea to go with non-standard evaluation. I'm a bit in favor of it here because this is meant to be a user-facing package, so if they have the tools available, that's all the better. My concern in this PR is that the internals rely on a long chain of dplyr verbs that might make things difficult to debug in the future. 

It would also be nice to describe in the documentation how the relative frequencies of the incubation frequencies are calculated. 

Also: when you are finished, don't forget to add your name to the DESCRIPTION file and items to NEWS.md

ffinger committed:
changes to make compatible with package requirements

ffinger commented:
Implemented all your changes.
It would probably be a good idea to add an interface that directly uses linelist epivars for data_of_onset and dates_of_exposure if they are defined.

ffinger committed:
added package suggestions to DESCRIPTION

ffinger committed:
add package imports to DESCRIPTION, assigning NULL to vars to avoid Note by R CMD check

ffinger committed:
adding tests

ffinger committed:
adding linelist to required packages for testign

ffinger commented:
@zkamvar any suggestions on how to test the catches for unexpected inputs? is a separate test for each catch needed?

ffinger committed:
catching constant incubation period

zkamvar commented on DESCRIPTION:
```diff
@@ -4,22 +4,35 @@ Version: 0.2.2
-             person("Dirk", "Schumacher", email = "mail@dirk-schumacher.net", role = "ctb")
-            ) 
+             person("Dirk", "Schumacher", email = "mail@dirk-schumacher.net", role = "ctb"),
+             person("Flavio", "Finger", email = "flavio.finger@lshtm.ac.uk", role = "ctb")
```
Since you are adding new functions, I would put yourself as an author

zkamvar commented on R/estimate_incubation.R:
```diff
@@ -0,0 +1,114 @@
+compute_incubation <- function(dates_exposure, date_onset){
+  z <- data.frame(date_onset = date_onset)
+  z$dates_exposure <- dates_exposure
+  incubation_period <- NULL #entry to avoid note by R CMD check
```
I'm not a big fan of this tactic. I would suggest to use this tactic:

```r
incubation_period <- quote(incubation_period)
weight <- quote(weight)

...

dplyr::select(!!incubation_period, !!weight)
```

Excruciating details here: https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html

zkamvar commented:
@ffinger, thank you for addressing the suggestions so quickly! I think you've done good work here 💯. The only other thing I would need from you is to put a little explanation of how things are calculated in the vignette and/or the documentation. I can take care of all the other minor details from there (if you are okay with that).

Misc comments/answers below

------------

> It would probably be a good idea to add an interface that directly uses linelist epivars for data_of_onset and dates_of_exposure if they are defined.

I think this is a good idea. We could have the user-specified columns as a backup. I'm also thinking of making linelist compatible with dplyr verbs, but I'm not exactly sure of how that can work. At the moment, there is the `lookup(df, "epivar")` function, but I need to put out some fires before I can get on that.

> any suggestions on how to test the catches for unexpected inputs? is a separate test for each catch needed?

Think about the number of possible ways something can go wrong and test each one individually. The more tests the better.

On another note, there is also the business of dealing with packages in "Suggests"... in short, CRAN gets upset when you assume that a user has these packages, so if you use them in examples, you need to wrap them in conditionals.

ffinger committed:
updated vignette and minor mods

ffinger commented:
@zkamvar thanks for the useful comments. I implemented them and added a little explanation in the vignette. Happy for you to take over from here.

zkamvar commented:
> @zkamvar thanks for the useful comments. I implemented them and added a little explanation in the vignette. Happy for you to take over from here.

Excellent! I'm currently at the end of my day, so I'll revisit this on Monday, but thank you again for contributing!

ffinger committed:
update vignette

ffinger committed:
making sure unobserved inc periods get rel freq 0

zkamvar commented:
Hi Flavio,

I'm going through the code and tweaking a few things to make sure it will pass CRAN's check when we eventually submit the new version and one thing struck me a bit odd: your function assumes a list column or a vector for exposures. I think it's a good idea to accommodate list columns if they exist, but I'm not sure that this will be the way the data would appear in the linelist. 

I can see a few possible ways of representing these data:

1. a single vector of dates 
2. two columns specifying the potential exposure range
3. a single column with the date range written in (e.g. `2 Jun - 5 Jun`) 
4. a single column with the potential exposure dates listed inside a single cell (read: a list column that doesn't know it's a list column yet)

I'll add a bit more documentation about the assumptions of the data and try to implement the case for \#2 today.

ffinger commented:
@zkamvar, the only reason why I decided to go with the list columns is that exposure dates don't necessarily need to be consecutive, giving a range doesn't work in those cases. I wanted to keep things as general as possible, but have thought about adding a second, simpler interface for giving ranges.

I don't really get the difference between a list column and your suggestion 4?

zkamvar committed:
clean up estimate_incubation

 - remove magrittr pipes
 - move magrittr to suggests
 - add comments to code
 - trade simple vector mutate calls to just assignments
 - fix up documentation

zkamvar commented:
> @zkamvar, the only reason why I decided to go with the list columns is that exposure dates don't necessarily need to be consecutive, giving a range doesn't work in those cases. I wanted to keep things as general as possible, but have thought about adding a second, simpler interface for giving ranges.

Agreed. It was a good move to start with this situation first since it's the most fine-grained. Given that there is the possibility for a third column being involved in this, how would you feel about switching the arguments of `empirical_incubation_dist()` to have `date_onset` first? This way, we can add a fourth argument that represents the end of the exposure range if that's what the linelist contains. 

> 
> I don't really get the difference between a list column and your suggestion 4?

I'm thinking of how linelists would be imported in the first place... if they had several dates in a single cell, they would be imported as character strings and would have to be coerced into list columns.

ffinger commented:
> Agreed. It was a good move to start with this situation first since it's the most fine-grained. Given that there is the possibility for a third column being involved in this, how would you feel about switching the arguments of empirical_incubation_dist() to have date_onset first? This way, we can add a fourth argument that represents the end of the exposure range if that's what the linelist contains

Yes, sounds good.

zkamvar committed:
update estimate incubation

 - add support for a start and stop column
 - add more catches for weird situations
 - TEST TEST TEST

zkamvar commented:
Allright, I've cleaned things up and added enough tests to make sure to get back up to 100% coverage. 

@ffinger, Thank you again for your contribution! 🎉 

In the future, it would be good to derive a small example with a known result as opposed to the roundabout way of sampling dates from a distribution and the comparing the result to an earlier saved result.
